Related links:

http://bugs.php.net/bug.php?id=41593
http://bugs.php.net/bug.php?id=36158
http://bugs.php.net/bug.php?id=43224
http://php-fpm.anight.org/

test script too:
<?php
echo php_sapi_name(), ' running ', PHP_VERSION, "<br>\n";
$i = 0;
while ($i < 35) {
    echo (++$i), "<br>\n";
    flush();
    sleep(1);
}
echo "end!<br>\n";
?>

diff -urNp -x '*.orig' php-5.2.17.org/sapi/cgi/cgi_main.c php-5.2.17/sapi/cgi/cgi_main.c
--- php-5.2.17.org/sapi/cgi/cgi_main.c	2021-10-23 19:05:04.013125101 +0200
+++ php-5.2.17/sapi/cgi/cgi_main.c	2021-10-23 19:05:06.359791768 +0200
@@ -100,6 +100,11 @@ static void (*php_php_import_environment
  */
 static int children = 0;
 
+#if PHP_FASTCGI
+/* Socket we are listening on incoming FastCGI connections */
+static int fcgi_fd = 0;
+#endif
+
 /**
  * Set to non-zero if we are the parent process
  */
@@ -1225,6 +1230,22 @@ void fastcgi_cleanup(int signal)
 }
 #endif
 
+#if PHP_FASTCGI
+/**
+ * Graceful shutdown. Close listening sockets.
+ */
+void fastcgi_graceful_shutdown(int signal)
+{
+#ifdef DEBUG_FASTCGI
+	fprintf(stderr, "FastCGI graceful shutdown, pid %d\n", getpid());
+#endif
+
+	/* Close the listening socket so new processes can reuse the same port */
+	closesocket(fcgi_fd);
+	fcgi_fd = 0;
+}
+#endif
+
 PHP_INI_BEGIN()
 	STD_PHP_INI_ENTRY("cgi.rfc2616_headers",     "0",  PHP_INI_ALL,    OnUpdateBool,   rfc2616_headers, php_cgi_globals_struct, php_cgi_globals)
 	STD_PHP_INI_ENTRY("cgi.nph",                 "0",  PHP_INI_ALL,    OnUpdateBool,   nph, php_cgi_globals_struct, php_cgi_globals)
@@ -1343,7 +1364,6 @@ int main(int argc, char *argv[])
 	int requests = 0;
 	int fastcgi = fcgi_is_fastcgi();
 	char *bindpath = NULL;
-	int fcgi_fd = 0;
 	fcgi_request request;
 	int repeats = 1;
 	int benchmark = 0;
@@ -1632,9 +1652,17 @@ consult the installation file that came
 					parent = 0;
 
 					/* don't catch our signals */
-					sigaction(SIGTERM, &old_term, 0);
 					sigaction(SIGQUIT, &old_quit, 0);
 					sigaction(SIGINT,  &old_int,  0);
+#if PHP_FASTCGI
+
+					/* call graceful shutdown handler for SIGTERM */
+					act.sa_flags = 0;
+					act.sa_handler = fastcgi_graceful_shutdown;
+					sigaction(SIGTERM, &act, &old_term);
+#else
+					sigaction(SIGTERM, &old_term, 0);
+#endif
 					break;
 				case -1:
 					perror("php (pre-forking)");
